Add `SourceKitObject` that represents `sourcekitd_object_t` in Swift

norio-nomura opened the pull request:
This adds `SourceKitObject`, `SourceKitObjectConvertible` and `UID` that salvaged from #274.
Fixes #489
This is a different solution than #488.
By using this changes, the sample code in https://github.com/jpsim/SourceKitten/issues/487#issue-294554672 can be written as following:
```swift
func requestSymbolSuperclasses(_ kind: SymbolKind, name: String, usr: String) -> [String: Any] {
    let response: [String: Any]

    do {
        response = try Request.customRequest(request: [
            "key.request": UID(RequestKind.symbolSuperclasses),
            "key.symbol": [
                "key.indexer.arg.symbol.name": name,
                "key.indexer.arg.symbol.kind": kind.rawValue,
                "key.indexer.arg.symbol.language": "Xcode.SourceCodeLanguage.Swift",
                "key.indexer.arg.symbol.resolution": usr
            ]
        ]).send()
    } catch {
        fatalError("Symbol superclasses request failed: \(error)")
    }

    return response
}
```
/cc: @ileitch

ileitch commented:
So much nicer! 💯

jpsim commented on Source/SourceKittenFramework/Request.swift:
```diff
@@ -196,7 +196,7 @@ public enum Request {
     /// A `cursorinfo` request for an offset in the given file, using the `arguments` given.
     case cursorInfo(file: String, offset: Int64, arguments: [String])
     /// A custom request by passing in the sourcekitd_object_t directly.
```
Should update this comment.

jpsim commented on Source/SourceKittenFramework/SourceKitObject.swift:
```diff
@@ -0,0 +1,132 @@
+
+import Foundation
+#if SWIFT_PACKAGE
+    import SourceKit
```
Minor: I generally avoid indenting compiler directive bodies. See https://github.com/jpsim/SourceKitten/blob/0.19.1/Source/sourcekitten/main.swift#L9-L15

jpsim commented on Source/SourceKittenFramework/UID.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+}
+
+extension UID: Hashable {
```
Thoughts on using the auto-synthesized `Hashable` implementation for Swift 4.1+?

```swift
extension UID: Hashable {
#if !swift(>=4.1)
    public var hashValue: Int {
        return uid.hashValue
    }

    public static func == (lhs: UID, rhs: UID) -> Bool {
        return lhs.uid == rhs.uid
    }
#endif
}
```

I don't feel strongly, but it seems like it might be easier to find code that can be removed once we require Swift 4.1+ to compile SourceKitten.

jpsim commented on Tests/SourceKittenFrameworkTests/SourceKitObjectTests.swift:
```diff
@@ -0,0 +1,38 @@
+
+class SourceKitObjectTests: XCTestCase {
+
+    func testExample() {
```
`testDescription`?

norio-nomura commented:
Thanks for reviews!
Sorry, I changed it considerably other than pointed out.
- Changed `SourceKitObject` to `class` for using `sourcekitd_request_release()` on `deinit`
- Change `SourceKitObjectConvertible.sourceKitObject` from a property to a throwing method
- Make `SourceKitObject` conforms to  `ExpressibleByBooleanLiteral`

Could you please review again?

norio-nomura commented:
@jpsim Should I open another PR against this branch for changes that added after previous reviews?

jpsim commented:
No need for a separate PR, I'll review this shortly. Looks mostly good, I think I just have a few pieces of small but breaking feedback

jpsim commented:
Sorry the delay is entirely my fault. I'll review & merge this now.

jpsim commented:
I renamed `SourceKitObjectConvertible.sourceKitObject` to `sourcekitdObject` to avoid confusion between 'SourceKitObject's and sourcekitd "objects". I hope that's ok with you. I'll merge this as soon as tests pass CI.

jpsim commented:
Tired of waiting for CI. Will fix if I get notified of failures when tests eventually run on master.

norio-nomura commented:
Some CI jobs for this PR failed since the branch has been removed before starting jobs.
I tried to re-run them, but it can't.
Since CI jobs for `master` branch passed, it may be ok.

jpsim commented:
Yes that’s expected.
